[Bug] Nondeterministic Assertions in KafkaAbstractSinkTest

### Search before reporting

- [x] I searched in the [issues](https://github.com/apache/pulsar/issues) and found nothing similar.


### Read release policy

- [x] I understand that [unsupported versions](https://pulsar.apache.org/contribute/release-policy/#supported-versions) don't get bug fixes. I will attempt to reproduce the issue on a supported version of Pulsar client and Pulsar broker.


### User environment

Broker version: 4.2.0-SNAPSHOT
Broker Operating system and hardware type: Linux x86_64 GNU/Linux
Broker Java version: 17.0.16

### Issue Description

The original test performed six assertions using the `expectThrows` method but was based on two incorrect assumptions. First, it implicitly assumed that field validation within [IOConfigUtils.java](https://github.com/apache/pulsar/blob/master/pulsar-io/common/src/main/java/org/apache/pulsar/io/common/IOConfigUtils.java) would occur in a consistent order. However, IOConfigUtils iterates over configuration fields using the loop `(Field field : getAllFields(clazz))`, and `getAllFields()` calls [Class.getDeclaredFields](https://docs.oracle.com/javase/8/docs/api/java/lang/Class.html#getDeclaredFields--), which returns an array of fields that are not sorted and not in any particular order. Second, the test initialized multiple required fields as null simultaneously, assuming the validation would fail in a specific order. In reality, the nondeterministic field iteration order can cause different missing fields to be detected first, leading to inconsistent exception messages and nondeterministic test failures due to different environments producing the contents in different orders despite the logical contents being the same. 
- pulsar-io/kafka/src/test/java/org/apache/pulsar/io/kafka/sink/KafkaAbstractSinkTest.java

I discovered the problem with the [NonDex](https://github.com/TestingResearchIllinois/NonDex) tool. NonDex systematically detects incorrect tests that rely on non-deterministic behaviors in Java APIs—like assuming order of name/value pairs in json files — by exploring all specification-allowed outcomes. It does this by instrumenting undetermined APIs and randomizing the returned order/behavior within what the spec allows. Failures it exposes reliably indicate flawed assumptions in order that were never promised and should be fixed.

### Error messages

```text
[INFO] -------------------------------------------------------
[INFO]  T E S T S
[INFO] -------------------------------------------------------
[INFO] Running org.apache.pulsar.io.kafka.sink.KafkaAbstractSinkTest
[ERROR] Tests run: 2, Failures: 1, Errors: 0, Skipped: 1, Time elapsed: 2.947 s <<< FAILURE! -- in org.apache.pulsar.io.kafka.sink.KafkaAbstractSinkTest
[ERROR] org.apache.pulsar.io.kafka.sink.KafkaAbstractSinkTest.testInvalidConfigWillThrownException -- Time elapsed: 0.004 s <<< FAILURE!
java.lang.AssertionError: expected [topic cannot be null] but found [acks cannot be null]
        at org.testng.Assert.fail(Assert.java:110)
        at org.testng.Assert.failNotEquals(Assert.java:1577)
        at org.testng.Assert.assertEqualsImpl(Assert.java:149)
        at org.testng.Assert.assertEquals(Assert.java:131)
        at org.testng.Assert.assertEquals(Assert.java:655)
        at org.testng.Assert.assertEquals(Assert.java:665)
        at org.apache.pulsar.io.kafka.sink.KafkaAbstractSinkTest.expectThrows(KafkaAbstractSinkTest.java:68)
        at org.apache.pulsar.io.kafka.sink.KafkaAbstractSinkTest.testInvalidConfigWillThrownException(KafkaAbstractSinkTest.java:202)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:569)
        at org.testng.internal.invokers.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:139)
        at org.testng.internal.invokers.InvokeMethodRunnable.runOne(InvokeMethodRunnable.java:47)
        at org.testng.internal.invokers.InvokeMethodRunnable.call(InvokeMethodRunnable.java:76)
        at org.testng.internal.invokers.InvokeMethodRunnable.call(InvokeMethodRunnable.java:11)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
        at java.base/java.lang.Thread.run(Thread.java:840)

[INFO] 
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR] org.apache.pulsar.io.kafka.sink.KafkaAbstractSinkTest.testInvalidConfigWillThrownException
[INFO]   Run 1: PASS
[ERROR]   Run 2: KafkaAbstractSinkTest.testInvalidConfigWillThrownException:202->expectThrows:68 expected [topic cannot be null] but found [acks cannot be null]                                                                                                                                            
[INFO] 
[INFO] 
[ERROR] Tests run: 1, Failures: 1, Errors: 0, Skipped: 0
```

### Reproducing the issue

You can replicate the problem with the [NonDex](https://github.com/TestingResearchIllinois/NonDex) tool command:
```
 mvn -pl pulsar-io/kafka   -Dtest=org.apache.pulsar.io.kafka.sink.KafkaAbstractSinkTest#testInvalidConfigWillThrownException -DforkCount=1  -DnondexRuns=1 -DnondexSeed=974622 -DreuseForks=false   edu.illinois:nondex-maven-plugin:2.2.1:nondex
```

### Additional information

_No response_

### Are you willing to submit a PR?

- [x] I'm willing to submit a PR!

--- a/pulsar-io/kafka/src/test/java/org/apache/pulsar/io/kafka/sink/KafkaAbstractSinkTest.java
+++ b/pulsar-io/kafka/src/test/java/org/apache/pulsar/io/kafka/sink/KafkaAbstractSinkTest.java
@@ -26,9 +26,12 @@ import java.io.IOException;
 import java.nio.ByteBuffer;
 import java.util.Collection;
 import java.util.HashMap;
+import java.util.List;
 import java.util.Map;
 import java.util.Properties;
 import java.util.concurrent.CompletableFuture;
+import java.util.function.Consumer;
+import java.util.function.Function;
 import org.apache.kafka.clients.producer.ProducerConfig;
 import org.apache.kafka.common.config.ConfigException;
 import org.apache.kafka.common.security.auth.SecurityProtocol;
@@ -65,7 +68,7 @@ public class KafkaAbstractSinkTest {
         } catch (Throwable e) {
             if (expectedType.isInstance(e)) {
                 T ex = expectedType.cast(e);
-                assertEquals(expectedMessage, ex.getMessage());
+                assertEquals(ex.getMessage(), expectedMessage);
                 return;
             }
             throw new AssertionError("Unexpected exception type, expected " + expectedType.getSimpleName()
@@ -74,10 +77,24 @@ public class KafkaAbstractSinkTest {
         throw new AssertionError("Expected exception");
     }
 
+    /**
+     * Creates a valid Kafka Sink configuration that is used by multiple test cases.
+     *
+     * @return a map containing all required Kafka sink configuration fields
+     */
+    private static Map<String, Object> validConfig() {
+        Map<String, Object> map = new HashMap<>();
+        map.put("bootstrapServers", "localhost:6667");
+        map.put("acks", "1");
+        map.put("topic", "topic_2");
+        map.put("batchSize", "16384");
+        map.put("maxRequestSize", "1048576");
+        return map;
+    }
+
     @Test
     public void testInvalidConfigWillThrownException() throws Exception {
         KafkaAbstractSink<String, byte[]> sink = new DummySink();
-        Map<String, Object> config = new HashMap<>();
         SinkContext sc = new SinkContext() {
             @Override
             public int getInstanceId() {
@@ -189,7 +206,7 @@ public class KafkaAbstractSinkTest {
 
             }
         };
-        ThrowingRunnable openAndClose = ()->{
+        Function<Map<String, Object>, ThrowingRunnable> runWith = config -> () -> {
             try {
                 sink.open(config, sc);
                 fail();
@@ -197,23 +214,54 @@ public class KafkaAbstractSinkTest {
                 sink.close();
             }
         };
-        expectThrows(IllegalArgumentException.class, "bootstrapServers cannot be null", openAndClose);
-        config.put("bootstrapServers", "localhost:6667");
-        expectThrows(IllegalArgumentException.class, "acks cannot be null", openAndClose);
-        config.put("acks", "1");
-        expectThrows(IllegalArgumentException.class, "topic cannot be null", openAndClose);
-        config.put("topic", "topic_2");
-        config.put("batchSize", "-1");
-        expectThrows(IllegalArgumentException.class, "Invalid Kafka Producer batchSize : -1", openAndClose);
-        config.put("batchSize", "16384");
-        config.put("maxRequestSize", "-1");
-        expectThrows(IllegalArgumentException.class, "Invalid Kafka Producer maxRequestSize : -1", openAndClose);
-        config.put("maxRequestSize", "1048576");
-        config.put("acks", "none");
-        expectThrows(ConfigException.class,
-                "Invalid value none for configuration acks: String must be one of: all, -1, 0, 1",
-                openAndClose);
-        config.put("acks", "1");
+
+        // Table of test cases for key removal and modification tests
+        record Case(
+                Consumer<Map<String, Object>> mutate,
+                Class<? extends Exception> expectedType,
+                String expectedMessage
+        ) {}
+
+        List<Case> cases = List.of(
+                // Missing bootstrapServers
+                new Case(config -> config.remove("bootstrapServers"),
+                        IllegalArgumentException.class,
+                        "bootstrapServers cannot be null"),
+
+                // Missing acks
+                new Case(config -> config.remove("acks"),
+                        IllegalArgumentException.class,
+                        "acks cannot be null"),
+
+                // Missing topic
+                new Case(config -> config.remove("topic"),
+                        IllegalArgumentException.class,
+                        "topic cannot be null"),
+
+                // Bad batchSize
+                new Case(config -> config.put("batchSize", "-1"),
+                        IllegalArgumentException.class,
+                        "Invalid Kafka Producer batchSize : -1"),
+
+                // Bad maxRequestSize
+                new Case(config -> config.put("maxRequestSize", "-1"),
+                        IllegalArgumentException.class,
+                        "Invalid Kafka Producer maxRequestSize : -1"),
+
+                // Invalid acks value
+                new Case(config -> config.put("acks", "none"),
+                        ConfigException.class,
+                        "Invalid value none for configuration acks: String must be one of: all, -1, 0, 1")
+        );
+
+        for (Case currCase : cases) {
+            var config = validConfig(); // set fresh, valid, baseline each time
+            currCase.mutate.accept(config); // remove or change one field
+            expectThrows(currCase.expectedType, currCase.expectedMessage, runWith.apply(config));
+        }
+
+        // Finally verify a valid config passes cleanly
+        var config = validConfig();
         sink.open(config, sc);
         sink.close();
     }

